File: libs/power-log-parser/src/lib/parsers/excavate-tier-changed-parser.ts
import { GameTag } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventProvider, GameEventHelper } from '../game-event';
import { Node, NodeType, PlayerEntity, Tag, TagChange } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class ExcavateTierChangedParser implements ActionParser {
	readonly ParserName = 'ExcavateTierChangedParser';

	private GameState: GameState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			node.Type === NodeType.TagChange &&
			(node.Object as TagChange).Name === (GameTag.CURRENT_EXCAVATE_TIER as number)
		);
	}

	AppliesOnCloseNode(node: Node, stateType: StateType): boolean {
		if (stateType !== StateType.PowerTaskList || node.Type !== NodeType.PlayerEntity) {
			return false;
		}
		const tags = (node.Object as PlayerEntity).Tags;
		return (
			tags.find((tag) => tag.Name === (GameTag.CURRENT_EXCAVATE_TIER as number)) != null ||
			tags.find((tag) => tag.Name === (GameTag.MAX_EXCAVATE_TIER as number)) != null
		);
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		const tagChange = node.Object as TagChange;
		const entity = this.GameState.CurrentEntities.get(tagChange.Entity);
		if (entity == null) {
			return null;
		}

		const maxExcavateTier = entity.GetTag(GameTag.MAX_EXCAVATE_TIER, 0);
		const currentExcavateTier = tagChange.Value;
		const cardId = entity.CardId;
		const controllerId = entity.GetEffectiveController();
		return [
			GameEventProvider.Create(
				tagChange.TimeStamp,
				'EXCAVATE_TIER_CHANGED',
				GameEventHelper.CreateProvider(
					'EXCAVATE_TIER_CHANGED',
					cardId,
					controllerId,
					entity.Id,
					this.StateFacade,
					{
						CurrentTier: currentExcavateTier,
						MaxTier: maxExcavateTier,
					},
				),
				true,
				node,
			),
		];
	}

	CreateGameEventProviderFromClose(node: Node): GameEventProvider[] | null {
		const playerEntity = node.Object as PlayerEntity;
		const maxExcavateTier =
			playerEntity.Tags.find((tag) => tag.Name === (GameTag.MAX_EXCAVATE_TIER as number))?.Value ?? 3;
		const currentExcavateTier =
			playerEntity.Tags.find((tag) => tag.Name === (GameTag.CURRENT_EXCAVATE_TIER as number))?.Value ?? 0;

		const controllerId = playerEntity.GetEffectiveController();
		return [
			GameEventProvider.Create(
				playerEntity.TimeStamp,
				'EXCAVATE_TIER_CHANGED',
				GameEventHelper.CreateProvider(
					'EXCAVATE_TIER_CHANGED',
					null as any,
					controllerId,
					playerEntity.Id,
					this.StateFacade,
					{
						CurrentTier: currentExcavateTier,
						MaxTier: maxExcavateTier,
					},
				),
				true,
				node,
			),
		];
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/ExcavateTierChangedParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;

namespace HearthstoneReplays.Events.Parsers
{
    public class ExcavateTierChangedParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public ExcavateTierChangedParser(ParserState ParserState, StateFacade facade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = facade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(TagChange)
                && (node.Object as TagChange).Name == (int)GameTag.CURRENT_EXCAVATE_TIER;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            List<Tag> tags = null;
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(PlayerEntity)
                && ((tags = (node.Object as PlayerEntity).Tags).Find(tag => tag.Name == (int)GameTag.CURRENT_EXCAVATE_TIER) != null
                    || tags.Find(tag => tag.Name == (int)GameTag.MAX_EXCAVATE_TIER) != null);
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            var tagChange = node.Object as TagChange;
            var entity = GameState.CurrentEntities.GetValueOrDefault(tagChange.Entity);
            if (entity == null)
            {
                return null;
            }

            var maxExcavateTier = entity.GetTag(GameTag.MAX_EXCAVATE_TIER, 0);
            var currentExcavateTier = tagChange.Value;
            var cardId = entity.CardId;
            var controllerId = entity.GetEffectiveController();
            return new List<GameEventProvider> { GameEventProvider.Create(
                tagChange.TimeStamp,
                 "EXCAVATE_TIER_CHANGED",
                GameEvent.CreateProvider(
                    "EXCAVATE_TIER_CHANGED",
                    cardId,
                    controllerId,
                    entity.Id,
                    StateFacade,
                    //null,
                    new {
                        CurrentTier = currentExcavateTier,
                        MaxTier = maxExcavateTier,
                    }),
                true,
                node) };
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            PlayerEntity playerEntity = node.Object as PlayerEntity;
            var maxExcavateTier = playerEntity.Tags
                .Find(tag => tag.Name == (int)GameTag.MAX_EXCAVATE_TIER)
                ?.Value ?? 3;
            var currentExcavateTier = playerEntity.Tags
                .Find(tag => tag.Name == (int)GameTag.CURRENT_EXCAVATE_TIER)
                ?.Value ?? 0;

            var controllerId = playerEntity.GetEffectiveController();
            return new List<GameEventProvider> { GameEventProvider.Create(
                playerEntity.TimeStamp,
                 "EXCAVATE_TIER_CHANGED",
                GameEvent.CreateProvider(
                    "EXCAVATE_TIER_CHANGED",
                    null,
                    controllerId,
                    playerEntity.Id,
                    StateFacade,
                    //null,
                    new {
                        CurrentTier = currentExcavateTier,
                        MaxTier = maxExcavateTier,
                    }),
                true,
                node) };
        }
    }
}
